enhancements/cguConditionEnhancement.md

```diff
@@ -0,0 +1,467 @@
+
+ ```yaml
+    Type:    "Progressing",
+    Status:  True/False,
+    Reason:  "MissingBlockingCR/IncompleteBlockingCR/MissingClusterLabel/InProgress",
+    Message: “”
+  ```
+
+  > Mohamed had a comment on this block: "In my opinion those are 2 different conditions 1) Initialized/Started means that the action (copying policies) is happening  2) Progressing/InProgress means that policy remediation started." To discuss.
+
+  6. The **UpgradeCompleted** and **UpgradeTimedOut** Reason can be moved to condition type: Succeeded
+
+  ```yaml
+    Type:    "Succeeded",
+    Status:  True/False,
+    Reason:  "UpgradeCompleted/UpgradeTimedOut",
+    Message: “”
+  ``` 
+
+### Workflow summary
+
+- When backup and Precaching is true:
+  * ClusterSelected -> Validated -> BackupSucceeded -> PrecacheSpecValid -> PrecachingSucceeded -> Progressing -> Succeeded
+ 
+
+- When backup and precaching is false:
+  * ClusterSelected -> Validated -> Progressing -> Succeeded
+
+### Mapping between Type, Status, Reason and Message:
+
+Based on the workflow that is described above, the summarized conditions,Reason and messages are grouped in a table below:
+
+if backup and/or precaching is false, the related types will not appear in `.status.conditions`.
+
+
+  | Type | Status| Reason| Message |
+  |------|-------|---------|--------|
+  ClustersSelected| True| Completed||
+  | |False | NotFound/NotPresent| `comment: mention the cluster list that is not part of the manangedCluster, must block the upgrade`|
+  Validated | True | Completed||
+  | | False | MissingManagedPolicy| `comment: must block the upgrade`|
+  | | False | InvalidPlatformImage| `comment: must block the upgrade`| 
+  `BackupSucceeded` | True | Completed | Backup is completed |
+  | | True | PartiallyDone | Failed to create backups for some clusters |
+  | | False | InProgress | Backup is in progress |
+  | | False | Failed | Backup failed for all the clusters |
+  `PrecacheSpecValid` | True | PrecacheSpecIsWellFormed/AsExpected | Pre-caching spec is valid and consistent|
+  | | False | NotAllManagedPoliciesExist| The ClusterGroupUpgrade CR has managed policies that are missing|
+  `PrecachingSucceeded` | True | Completed | Precaching is completed |
+  | | True | PartiallyDone | Failed to create preCaching for some clusters | 
+  | | False | InProgress | Precaching is in progress | 
+  | | False | Failed | PreCaching failed for all the clusters |
+  Progressing| True | InProgress||
+  | | False | Completed | |
+  | | False | NotEnabled| The ClusterGroupUpgrade CR is not enabled |
+  | | False | MissingBlockingCR | `comment: message must contain which blocking CRs are missing`|
+  | | False | IncompleteBlockingCR | `comment: blocked by other CR which is not completed, mention which CR is blocking`| 
+  Succeeded| True | UpgradeCompleted||
+  | | False | UpgradeTimedOut | |
+
+
+- If `ClustersSelected` is false, it should block the subsequent states, meaning we do not proceed backup, precaching, upgrade 
```

Comment:
ideally, if a CR is not enabled we should have below flow:
 
` ClusterSelected -> Validated ->  Progressing `

The only thing that i'm not sure is if the CR isn't enabled, do we proceed to Precaching if it is set to true because it takes a lot of time and probably it doesn't need to be done just before the upgrade. In that case, we can have this below workflow:
`ClusterSelected -> Validated -> PrecacheSpecValid -> PrecachingSucceeded -> Progressing`

But for backup, it needs to be done before starting of an upgrade. Any thoughts? @jc-rh @vitus133

Reply:
My understanding is that all prep/non disruptive work should be done prior to enabling the CGU. This is to minimize the time requirement on the maintenance window.

Reply:
Since the backup takes in the range of 30s-300s (depends on the amount of workload running in SNO and the acm api overload condition), it is a very short lived task. The idea is to take the backup just before an upgrade is to have the most recent cluster and workload backup rather than  having a stale one.

Reply:
Pre-caching should be done as soon as the CR is applied.
Backup should be started when the CR is enabled.

Reply:
After the discussion, i think it is reasonable to have below workflow if enabled is false:

`ClusterSelected -> Validated -> BackupSucceeded -> PrecacheSpecValid -> PrecachingSucceeded -> Progressing`

But in any case, in the documentation, we must indicate that, if there is some modification done in the SNO after creating the CGU, re-create the CGU if changes were made in the SNOs
@fontivan

Reply:
Hi @donpenney, tagging you here. As Brent mentioned, Pre-caching may be done before Backup can start in case enabled is false, how do we skip those precaching pods from the backup or if not how do we make sure the precaching job doesn't start again after the recovery? One solution i was thinking is to cleanup the precaching pods once it is done successfully. wdyt?

Reply:
@fontivan This impacts what you are implementing.

Reply:
Another point to consider. Currently we build the remediation plan based on pre-caching/backup results before the enable. If backup is done after enable, the plan needs to be built later as well, or get revised. This would be a behavior change from a user's point of view. They can no longer preview the final plan before enabling the CGU.